### DISCUSSION TITLE
How to set reference value

### DISCUSSION THREAD

**Discussion by hui-zhou-a** [ansys/pyoptislang#153] [Q&A] [answered]
Hi experts, 

I would like to modify the reference value of a parameter, but without success. The test code is as below.

```
from pathlib import Path

from ansys.optislang.core import Optislang


osl = Optislang(project_path=Path("osl_parametric.opf"))
rs = osl.project.root_system
op = rs.get_nodes()[2]
op.parameter_manager.get_parameters()[0].reference_value = 1
assert op.parameter_manager.get_parameters()[0].reference_value == 1
```

Thanks in advance.
Hui

**Comment by PanekOndrej** [comment, marked as answer]
Hi @ring630 , 

this is not currently possible directly through `Parameter` object, since it's not connected to the server nor shares the same address space with optiSLang project, it's data class that stores copy parameter's properties.  

To explain what happened: 

In following line of code you queried server to obtain first parameter's data and modified reference value in your local copy. 

`op.parameter_manager.get_parameters()[0].reference_value = 1`

Then you queried server again to obtain first parameter's reference value ( ... compared with 1 and used assert). 

`assert op.parameter_manager.get_parameters()[0].reference_value == 1 `


Development of pyOptiSLang is still in progress, we are working on suitable functionality for your task.

The best way, how to achieve your goal right now is through ``TcpOslServer``, that extends available functionality.


```
from ansys.optislang.core import Optislang 
from ansys.optislang.core.tcp_osl_server import TcpOslServer 
import ansys.optislang.core.server_commands as sc 
from pathlib import Path 

osl = Optislang(project_path=Path("osl_parametric.opf")) 
tcp: TcpOslServer = osl.get_osl_server() 
rs = osl.project.root_system 

# please note that method rs.find_nodes_by_name(name='<inser your nodes name>')[0] may be used to find desired node 
# instead of following command 

op = rs.get_nodes()[2] 
properties = op.get_properties() 
parameter_manager_dict = properties['properties']['ParameterManager'] 

# insert your code to modify parameter inside dictionary
# probably something like:  
# parameter_manager_dict['parameter_container'][0]['reference_value'] = 1

# this will update your nodes property 
tcp.send_command(sc.set_actor_property(op.uid, 'ParameterManager', parameter_manager_dict)) 

# dispose osl if you're finished 
osl.dispose() 
```

Hopefully it helps, best regards, 

Ondrej
